format/index.md

```diff
@@ -812,6 +821,23 @@ include additional [fields] in the response.
 GET /articles?include=author&fields[articles]=title,body&fields[people]=name
 ```
 
+If both query parameters `include` and `fields[TYPE]` are provided by the client, 
+values of `TYPE` **MUST** either refer to the primary data or to resources 
+returned in `include`, otherwise the server **MUST** return `400 Bad Request`.
+
+If all of the following are true:
+
+- the client provides `include` parameter, and the resource of type `A` is referred to in it,
+- the client provides parameter `fields[B]`, requiring sparse fieldsets for the type `B`,
+- resource of type `A` is actually referenced by type `B`, but 
+reference to it is not requested in `fields[B]`, 
+
+the server **MUST** return `400 Bad Request`.
+
+> Note: In two clauses preceding this note, "refer" is assumed to correctly track 
+heterogeneous relationships, where the type of the resource being referred to 
+in the relationship can be different from the name of the relationship itself.
+
```

Comment:
**1st clause**

So, this part deals with `fields[TYPE]` when `TYPE` is not applicable to the response.

Why is this a problem to begin with? Why does the server need to complain about a redundant query param? And since a heterogeneous collection may or may not contain items of a particular type, the validity of the request would have to be contingent on what types *might* theoretically be present in the response. This is just too burdensome, especially considering that this restriction doesn't seem to have any tangible utility.

**2nd clause**

I know what this PR is about, yet after a few readings I still don't understand. Perhaps I've been hit by early-onset dementia, but I would like us to find some ways to simplify this whole ordeal.

**The Note**

Relationship names and heterogeneity are two independent properties, so again I'm not following.